_posts/2015-04-17-Enums-match-mutation-and-moves.md

```diff
@@ -0,0 +1,636 @@
+        Answer::Bingo  => println!("we won with {}!", prior_guess),
+    }
+}
+
+#[test]
+fn demo_suggest_guess() {
+	suggest_guess(10, Answer::Higher);
+	suggest_guess(20, Answer::Lower);
+	suggest_guess(19, Answer::Bingo);
+}
+```
+
+(Incidentally, nearly all the code in this post is directly
+executable; you can cut-and-paste the code snippets into a file
+`demo.rs`, compile the file with `--test`, and run the resulting
+binary to see the tests run.)
+
+Patterns can also match structured data (e.g. tuples, slices, user-defined
+data types) via corresponding patterns. In such patterns, one often
+binds substructure of the input to local variables (identifier patterns),
+for use either in the arm's predicate or in its result.
+
+The special `_` pattern matches any single value, and is often used as
+a catch-all; the special `..` pattern generalizes this by matching any
+*series* of values or name/value pairs.
+
+Also, one can collapse multiple patterns into one arm by separating the
+patterns by vertical bars (`|`); thus that arm matches either this pattern,
+or that pattern, et cetera.
+
+These features are illustrated in the following revision to the
+guessing-game answer generation strategy:
+
+```rust
+struct GuessState {
+    guess: u32,
+    answer: Answer,
+    low: u32,
+    high: u32,
+}
+
+fn suggest_guess_smarter(s: GuessState) {
+    match s {
+        GuessState { answer: Answer::Bingo, guess: p, .. } => {
+            println!("we won with {}!", p);
+        }
+        GuessState { answer: Answer::Higher, low: _, guess: l, high: h } |
+        GuessState { answer: Answer::Lower,  low: l, guess: h, high: _ } => {
+            let mid = l + ((h - l) / 2);
+            println!("lets try {} next", mid);
+        }
+    }
+}
+
+#[test]
+fn demo_guess_state() {
+	suggest_guess_smarter(GuessState {
+		guess: 20, answer: Answer::Lower, low: 10, high: 1000
+	});
+}
+```
+
+This ability to simultaneously perform case analysis *and* bind input
+substructure leads to powerful, clear, and concise code, focusing the
+reader's attention directly on the data relevant to the case at hand.
+
+That is `match` in a nutshell.
+
+So, what is the interplay between this construct and Rust's approach to
+ownership and safety in general?
+
+### Exhaustive case analysis
+
+> ...when you have eliminated all which is impossible,
+> then whatever remains, however improbable, must be the truth.
+>
+> -- Sherlock Holmes (Arthur Conan Doyle, "The Blanched Soldier")
+
+`match` enforces exhaustive case analysis, which helps catch bugs in
+program logic and ensures that the value of a match expression is
+well-defined.
+
+Case analysis is an important problem solving technique: Dividing a
+problem into some number of separate cases, and then analyzing each
+case individually.
+
+For this method of problem solving to work, the cases must be
+*collectively exhaustive*; otherwise, a case that was not covered
```

Comment:
I think somewhere here we should define the term *exhaustive*, just to say something like "Rust makes sure you have a match arm that covers every possible variant"